autorest/adal/token.go

```diff
@@ -941,9 +941,12 @@ func (spt *ServicePrincipalToken) refreshInternal(ctx context.Context, resource
 	}
 
 	var resp *http.Response
-	if isMSIEndpoint(spt.inner.OauthConfig.TokenEndpoint) && !MSIAvailable(ctx, spt.sender) {
-		// return a TokenRefreshError here so that we don't keep retrying
-		return newTokenRefreshError("the MSI endpoint is not available", nil)
+	if isMSIEndpoint(spt.inner.OauthConfig.TokenEndpoint) {
+		resp, err = getMSIEndpoint(ctx, spt.sender)
+		if err != nil {
+			// return a TokenRefreshError here so that we don't keep retrying
+			return newTokenRefreshError(fmt.Sprintf("the MSI endpoint is not available. Status Code = '%d'. Failed HTTP request to MSI endpoint: %v", resp.StatusCode, err), resp)
```

Comment:
If `err` is not `nil` it's not safe to dereference `resp` here (there very well might be no response).

Reply:
better? added a unit test

Reply:
it definitely fixes the panic.  does it add anything of value though?  from the docs it looks like there's only a very specific case where `Do()` will return a non-nil response and error
```go
// On error, any Response can be ignored. A non-nil Response with a
// non-nil error only occurs when CheckRedirect fails, and even then
// the returned Response.Body is already closed.
```
I'm fine with leaving the change as-is but wanted to point this out.

Out of curiosity, what prompted this change?

Reply:
> Out of curiosity, what prompted this change?

we find our service (AKS) hitting this error path frequently after switching over to heavy use of MSI. I suspect the responses will be 429s, which appear to be failing nil check I'm enriching. I assumed the full response would be attached too in this case, apologies for not being more familiar with the sender semantics. Maybe you could suggest a better behavior?

Reply:
I guess I should probably flip this -- we should return the error if not nil, and check the http status code otherwise, and do logic similar to inside `retryForIMDS`. But the intent of this code is to fail fast, so maybe that's not really worthwhile?

As an end user, I'll probably end up wrapping the refresh logic myself with backoff to avoid this shortcircuit logic.

Reply:
Per the docs, a non-2xx status code doesn't return an error.  The only time I'd expect `Do()` to return an error is if the underlying transport can't connect to the MSI endpoint.  Ideally this should only happen if the MSI endpoint truly isn't available (either it's not enabled on your VM or you're running on a dev box).  If you're seeing this return `false` when it should return `true` I'd like to fix this if possible.

Reply:
(I guess that also means we weren't getting throttled, but had some transient connectivity failures? interesting to know)

Reply:
> is if the underlying transport can't connect to the MSI endpoint. Ideally this should only happen if the MSI endpoint truly isn't available

we see on the order of 10k successful calls and single digits of failure over long enough time periods (hours). MSI is definitely there so after perusing the docs I agree it's probably network connectivity failures. It'd be nice to get the errors, though 😃 

let me make a few additional tweaks, perhaps.

Reply:
maybe more like that?

Reply:
This works, and should help add context to the failure so we can make the implementation more robust.